toothpick-runtime/src/main/java/toothpick/ToothPick.java

```diff
@@ -52,14 +57,17 @@ public static Scope openScopes(Object... names) {
    * Otherwise a new scope is created.
    */
   public static Scope openScope(Object name) {
-    synchronized (ToothPick.class) {
-      Scope scope = MAP_KEY_TO_SCOPE.get(name);
-      if (scope == null) {
-        scope = new ScopeImpl(name);
-        MAP_KEY_TO_SCOPE.put(name, scope);
-      }
+    Scope scope = MAP_KEY_TO_SCOPE.get(name);
+    if (scope != null) {
```

Comment:
yes, t2 should get something. The idea of the lock free structures is that when a thread reads and another thread  writes, the reader sees the data structure either before or after the change, but not during the transition. And I believe it's the case here. In a multi-threaded context, you should be careful that your scope was not closed before using it. 

Closing a scope will detach it from the tree, and C will become a root of a new tree indeed. With synchronized, the exact same problem will happen a millisecond later if some thread closes the scope you just opened.

My point of view is that this can be more or less adressed by DI itself. In Android for instance, the main thread starts a background threads with a callback, when the main thread leaves the context, it should remove the callback and interrupt the background thread or not. If the background thread uses only the application level scope, it can be allowed to continue will the activity is dead, but if it uses activity scope or lower, it should be stopped. 

So in one case, you can use the scope, in the other not. Really, I don't think this can be fully solved by DI. What I favored here is performance.